src/main/java/com/jcraft/jsch/OpenSshCertificateAwareIdentityFile.java

```diff
@@ -0,0 +1,285 @@
+
+
+  /**
+   * Creates a new certificate-aware identity from file paths.
+   *
+   * @param prvfile path to the private key file
+   * @param pubfile path to the certificate file
+   * @param instLogger logger instance for debugging
+   * @return a new Identity instance
+   * @throws JSchException if the files cannot be loaded or parsed
+   */
+  static Identity newInstance(String prvfile, String pubfile, JSch.InstanceLogger instLogger)
+      throws JSchException {
+    byte[] prvkey;
+    byte[] certificateFileContent;
+
+    try {
+      prvkey = Util.fromFile(prvfile);
+      certificateFileContent = Util.fromFile(pubfile);
+    } catch (IOException e) {
+      throw new JSchException(e.toString(), e);
+    }
+    return newInstance(prvfile, prvkey, certificateFileContent, instLogger);
+  }
+
+  /**
+   * Creates a new certificate-aware identity from byte arrays.
+   *
+   * @param name the identity name
+   * @param prvkey the private key bytes
+   * @param certificateFileContentBytes the certificate bytes
+   * @param instLogger logger instance for debugging
+   * @return a new Identity instance
+   * @throws JSchException if the certificate cannot be parsed
+   */
+  static Identity newInstance(String name, byte[] prvkey, byte[] certificateFileContentBytes,
+      JSch.InstanceLogger instLogger) throws JSchException {
+    String certificateFileContentString = new String(certificateFileContentBytes, UTF_8);
+    OpenSshCertificate cert;
+    byte[] certPublicKey;
+    KeyPair kpair;
+    String declaredKeyType;
+    String comment;
+    String base64KeyData;
+
+
+    try {
+      declaredKeyType = extractKeyType(certificateFileContentString);
+      base64KeyData = extractKeyData(certificateFileContentString);
+      comment = extractComment(certificateFileContentString);
+      byte[] keyData =
+          fromBase64(base64KeyData.getBytes(UTF_8), 0, base64KeyData.getBytes(UTF_8).length);
+      cert = OpenSshCertificateParser.parse(instLogger, keyData);
+
+      // keyType
+      if (cert.getKeyType().isEmpty() || !cert.getKeyType().equals(declaredKeyType)) {
+        instLogger.getLogger().log(Logger.WARN,
+            "Key type declared at the beginning of the certificate file, does not correspond to the encoded key type. Declared type: '"
+                + declaredKeyType + "' - Encoded Key type: '" + base64KeyData + "'"
+                + cert.getKeyType() + "'");
+      }
+
+      if (!cert.isValidNow()) {
+        instLogger.getLogger().log(Logger.WARN,
+            "certificate is not valid. Valid after: " + toDateString(cert.getValidAfter())
+                + " - Valid before: " + toDateString(cert.getValidBefore()));
+      }
+
+      certPublicKey = cert.getCertificatePublicKey();
+      kpair = KeyPair.load(instLogger, prvkey, certPublicKey);
+
+
+    } catch (Exception e) {
+      throw new JSchException(e.toString(), e);
+    }
+    return new OpenSshCertificateAwareIdentityFile(name, declaredKeyType, base64KeyData, cert,
```

Comment:
- Should the cert be checked to confirm that it is a `SSH2_CERT_TYPE_USER` type cert?
- Should the cert be checked to confirm that it contains one or more principals?

Reply:
The server is responsible for strictly validating the certificate type (i.e., that it's SSH2_CERT_TYPE_USER) and confirming the presence of principals. The client's role here is to present the certificate, so adding that validation on the client side isn't strictly necessary.

Reply:
What is the behavior of OpenSSH client when you attempt to use an `SSH2_CERT_TYPE_HOST` for authentication?
I.e.:
```shell
ssh -o CertificateFile=a_host_type_cert.pub -i the_matching_private_key w.x.y.z
```

Does the OpenSSH client (`ssh`) proceed to use the host type certificate, and then the server rejects the auth request?
Or does the OpenSSH client (`ssh`) refuse to even try to perform the connection (since the user specified the wrong type of cert)?

Reply:
Ok, thanks for confirming!